shell/browser/protocol_registry.cc

```diff
@@ -21,6 +55,12 @@ ProtocolRegistry::~ProtocolRegistry() = default;
 
 void ProtocolRegistry::RegisterURLLoaderFactories(
     content::ContentBrowserClient::NonNetworkURLLoaderFactoryMap* factories) {
+  // Override the default FileURLLoaderFactory to support asar archives.
+  //
+  // Note that |insert| or |emplace| should not be used here, as they do nothing
+  // if the key already exists, while we would like to overwrite the factory.
+  (*factories)[url::kFileScheme] = std::make_unique<AsarURLLoaderFactory>();
```

Comment:
When extension system is enabled, it also registers a factory for `file://` https://github.com/electron/electron/blob/master/shell/browser/electron_browser_client.cc#L1281 , it seems to override this one for sub resources. Will that be okay ?

Reply:
Actually the problem is other way around, the extension system uses `emplace` to register, so it won't override our factory. But then question, its trying to register a file provider with stricter rules https://github.com/electron/electron/blob/master/shell/browser/electron_browser_client.cc#L1202 and that won't be available, is that okay ? Or can we accommodate that configuration here ?

Reply:
The url factory installed by extension system is to support `file://` under `chrome-extension://` schemes, Chromium only installs file url factory for `file://` and `about:blank` so it is fine for extension system to use `emplace`.

The problem is we are unconditionally installing asar url factory for all protocols, I have updated the code to only install it when needed.